terraform/addons/byo-s3-logging-destination-firehose/firehose/main.tf

```diff
@@ -0,0 +1,119 @@
+data "aws_region" "current" {}
+data "aws_caller_identity" "current" {}
+
+data "aws_iam_policy_document" "osquery_firehose_assume_role" {
+  statement {
+    effect  = "Allow"
+    actions = ["sts:AssumeRole"]
+    principals {
+      identifiers = ["firehose.amazonaws.com"]
+      type        = "Service"
+    }
+  }
+}
+
+data "aws_iam_policy_document" "firehose_policy" {
+  statement {
+    effect = "Allow"
+    actions = [
+      "s3:AbortMultipartUpload",
+      "s3:GetBucketLocation",
+      "s3:GetObject",
+      "s3:ListBucket",
+      "s3:ListBucketMultipartUploads",
+      "s3:PutObject",
+      "s3:PutObjectAcl" // required according to https://docs.aws.amazon.com/firehose/latest/dev/controlling-access.html#using-iam-s3
+    ]
+    resources = [
+      "arn:aws:s3:::${var.results_destination_s3_bucket}",
+      "arn:aws:s3:::${var.results_destination_s3_bucket}/*",
+      "arn:aws:s3:::${var.status_destination_s3_bucket}",
+      "arn:aws:s3:::${var.status_destination_s3_bucket}/*"
+    ]
+  }
+
+  statement {
+    effect    = "Allow"
+    actions   = ["kms:GenerateDataKey*"]
+    resources = [var.kms_key_arn]
+  }
+
+  statement {
+    effect    = "Allow"
+    actions   = ["logs:PutLogEvents"]
+    resources = [
+      "arn:aws:logs:${data.aws_region.current.id}:${data.aws_caller_identity.current.account_id}:log-group:/aws/kinesisfirehose/${var.firehose_results_name}:*",
+      "arn:aws:logs:${data.aws_region.current.id}:${data.aws_caller_identity.current.account_id}:log-group:/aws/kinesisfirehose/${var.firehose_status_name}:*"
+    ]
+  }
+}
+
+resource "aws_iam_role" "firehose" {
```

Comment:
I think we can get around the "chicken-and-the-egg" problem of requiring the dummy values by having the instructions to be apply this side with just a `-target` of this resource and then the ARN would be available to provide to the "other side".  Then we would have the bucket and kms id/arns to be able to do a full apply.

Then we wouldn't need dummy "change me" roles at all hopefully.

Reply:
I talked with @edwardsb about this. We actually don't need that if we use a predictable naming convention. When customer Foobar wants Fleet Cloud, we will know what we want to name the IAM role for this, so we give them a .tfvars with the role name set. AWS doesn't care that the role hasn't been made yet though, so once they report back with the customer side applied and the values we need that are random (such as KMS key ARN), then we can apply our side without targeting.

Reply:
I think the way @zwinnerman-fleetdm and I figured is that we'll have a role that is deterministic, say "<customer>-firehose-role".

That way when they go to apply, we just tell them what the tf var should be, and when we apply on our side, it'll all just match up.